models/multiphase/d2q9_pf_velocity/Dynamics.c.Rt

```diff
@@ -114,8 +114,27 @@ CudaDeviceFunction vector_t calcGradPhi(){
 			gradPhi.x = (PhaseF(1,0) - PhaseF(-1,0))/3.0 + (PhaseF(1,-1) - PhaseF(-1,-1) )/6.0;
 			gradPhi.y = 0.0;
 		} else if ((NodeType & NODE_BOUNDARY) == NODE_Wall){
-			gradPhi.x = (PhaseF(1,0) - PhaseF(-1,0))/3.0 + (PhaseF(1,1) - PhaseF(-1,-1) + PhaseF(1,-1) - PhaseF(-1,1))/12.0;
-			gradPhi.y = (PhaseF(0,1) - PhaseF(0,-1))/3.0 + (PhaseF(1,1) - PhaseF(-1,-1) + PhaseF(-1,1) - PhaseF(1,-1))/12.0;
+			// compute gradient with one-sided diffs but first check if neighbours are nans
+			bool nan1 = isnan(PhaseF(1,0));
+			bool nan2 = isnan(PhaseF(-1,0));
+			bool nan3 = isnan(PhaseF(0,1));
+			bool nan4 = isnan(PhaseF(0,-1));
+
+			if ((nan1 && nan2) || (!nan1 && !nan2)){
```

Comment:
This 1. doesn't seem consistent with the previous implementation 2. gradient has different sign depending on situation 3. PhaseF are read multiple times.

I think something very consistent and robust could be achieved by some kind of a conditional average like:
```c++
gradPhi.x = 0;
gradPhi.y = 0;
phi_a = PhaseF(0,0);
for (/* all directions */) {
  phi_b = PhaseF(/* direction */);
  if (!isnan(phi_b)) {
    gradPhi.x += (phi_a - phi_b)*[/* weight */]*[/* direction x */];
    gradPhi.y += (phi_a - phi_b)*[/* weight */]*[/* direction y */];
    denom += [/* weight */];
  }
  gradPhi.x = gradPhi.x / denom;
}
gradPhi.y = gradPhi.y / denom;
```

Reply:
Thank you. This is a nicer method than attempting strange 1-sided diffs based on checking neighbours. The implementation is now:
`	

	gradPhi.x = 0.0;
	gradPhi.y = 0.0;

	real_t phi_a = PhaseF(0,0);
	real_t phi_b, denom = 0.0;

	for (int i = 1; i<9; i++) {
		phi_b = PhaseF_dyn(d2q9_ex[i],d2q9_ey[i]);
		if (!isnan(phi_b)){
			gradPhi.x += wf[i]*d2q9_ex[i]*(phi_b-phi_a);
			gradPhi.y += wf[i]*d2q9_ey[i]*(phi_b-phi_a);
			denom += wf[i];
		}
	}
	gradPhi.x = gradPhi.x/denom/cs2;
	gradPhi.y = gradPhi.y/denom/cs2;

`

I will do some more rigorous testing of results tomorrow and add a similar implementation for the d3q27 model :)